### DISCUSSION TITLE
understanding active_matrix_from_extrinsic_euler_...

### DISCUSSION THREAD

**Discussion by romainVala** [dfki-ric/pytransform3d#169] [Q&A] [answered]
Hello

First thanks for providing this code, and the very good documentation. Finding the correct euler convention for a given software can become messy. I really appriciate "Transformation Ambiguities and Conventions" but I may not sure to fully understand it yet. (so I may be wrong)

I just want to look at the difference between (for instance) those two transforms
active_matrix_from_extrinsic_euler_xyz
active_matrix_from_extrinsic_euler_yxz

I agree with the implementation of the xyz one, but not of the yxz one.
I expect to both have the same basis, but just the order of multiplication should change. ( in your implementation you change the order of multiplication but also the basis ...

To make it coherent  I need to change your implementation:
instead of 
```python
    R = active_matrix_from_angle(2, gamma).dot(
        active_matrix_from_angle(0, beta)).dot(
        active_matrix_from_angle(1, alpha))
```
by
```python
    R = active_matrix_from_angle(2, gamma).dot(
        active_matrix_from_angle(0, alpha)).dot(
        active_matrix_from_angle(1, beta)
```

does it make sense or I misinterpreted the conventions ... ?

Many thanks

**Comment by AlexanderFabisch** [comment, marked as answer]
I think I don't really understand the question. I guess by basis you mean the axis of rotation? 0 -> x, 1, -> y, 2 -> z?

The order of concatenation of the matrices is determined by the order of rotation axes (xyz, yxz, ...) and whether it is an extrinsic or an intrinsic convention.

The yxz convention means that the first angle (alpha) rotates around the y-axis, the second angle (beta) rotations around the x-axis, and the third angle (gamma) rotates around the z-axis. When the extrinsic convention is used, you concatenate rotation matrices from right to left.

I don't know why you would want to change the order of arguments.

**Reply by romainVala** [reply]
I want to use your euler to matrix (and matrix to euler) functions with the 2 possible convention of simpleitk software
if you can take a quick look at ziwy answer there https://discourse.itk.org/t/understanding-sitk-euler3dtransform-convention/4422

In short they define, Rx Ry Rz, as your active_matrix_from_angle, and 
their "ZYX" convention is then
Rz.Ry.Rx       

and their "ZYX" convention is 
Rz.Rx.Ry

I do obtain the exact same affine as itk ('ZYX' convention ) when I use your active_matrix_from_extrinsic_euler_xyz 

But if I want to reproduce the 'ZXY' convention I need to do the change in active_matrix_from_extrinsic_euler_yxz

**Reply by AlexanderFabisch** [reply]
Oh, that convention is really confusing to me. I think the reason for this is the strange API. You first pass the angles, translate them to rotation matrices in the order xyz, and then define in which order the transformation matrices are concatenated. That feels completely wrong to me.

If you want to get the same result, just change the order in which you pass the angles to `active_matrix_from_extrinsic_euler_yxz`:

```python
active_matrix_from_extrinsic_euler_yxz([beta, alpha, gamma])
```

I still find it weird, because it makes much more sense that the first angle actually rotates around the first axis, second angle around the second axis, and the third angle around the third axis of the convention that you actually use.

**Reply by romainVala** [reply]
thanks for your help, it indeed works to swap alpha and beta with you _yxz transform.

does it mean that it is the simpleitk interpretation of eurler angle that is wrong ? (it is a well know software in our field, so I find it strange ...)

**Reply by AlexanderFabisch** [reply]
Not wrong per se, but I find it very confusing. Which field is it if I may ask?

**Reply by romainVala** [reply]
I work with MRI data, but itk (simpleitk ) is dedicated to 3D (medical) Image processing (based on the itk C++ library)
